### PR TITLE
Allow declaring enum in external table structure

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Allow declaring enum in external table structure

Currently this query doesn't work:
```bash
curl -s "${http_url}temp_structure=x+Enum8('foo'%3D1,'bar'%3D2),y+Int" -F "$(printf 'temp='foo'\t1');filename=data1" -F "query=SELECT * FROM temp"
```

Because `ExternalTable` uses `splitInto<' ', ','>` to parse column list and confuses `,` in enum declaration.
The solution in this PR is to use proper parser to parse the external structure first before using `splitInto<' ', ','>`.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by canhld94** [comment]
Tagging the last people to touch `ExternalTable.cpp` code @alexey-milovidov

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 112c05c017ccbefb4a672550015b6a21207e5e9b with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57857/112c05c017ccbefb4a672550015b6a21207e5e9b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7216885587">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57857/112c05c017ccbefb4a672550015b6a21207e5e9b/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Core/ExternalTable.cpp [hunk 1]
```diff
@@ -55,23 +70,61 @@ void BaseExternalTable::clear()
...
+            else
+            {
+                structure.clear();
+                parse_structure_with_parser = false;
+                break;
+            }
+        }
+    }
+
+    if (!parse_structure_with_parser)
```
Comment (alexey-milovidov) [inline src/Core/ExternalTable.cpp:95]: I see this fallback as unneeded. The previous code was simply wrong. Splitting by comma is always the wrong way to parse something. Your code is better.

File: src/Core/ExternalTable.cpp [hunk 2]
```diff
@@ -55,23 +70,61 @@ void BaseExternalTable::clear()
...
+    ASTPtr type_list_raw = tryParseQuery(parser, pos, pos+argument.size(), error, false, "", false, 0, 0);
+    if (type_list_raw != nullptr)
+    {
+        for (size_t i = 0; i < type_list_raw->children.size(); ++i)
+            structure.emplace_back("_" + toString(i + 1), type_list_raw->children[i]->getColumnNameWithoutAlias());
+    }
+    else
+    {
+        std::vector<std::string> vals;
+        splitInto<' ', ','>(vals, argument, true);
```
Comment (alexey-milovidov) [inline src/Core/ExternalTable.cpp:123]: .

**Review by alexey-milovidov — APPROVED** [review]
Everything else LGTM.

File: src/Core/ExternalTable.cpp [hunk 3]
```diff
@@ -55,23 +70,61 @@ void BaseExternalTable::clear()
...
 {
-    std::vector<std::string> vals;
-    splitInto<' ', ','>(vals, argument, true);
+    /// First try to parse table structure with `ParserNameTypePairList`, this allows user to declare Enum types in the structure
+    ParserNameTypePairList parser;
+    const auto * pos = argument.data();
+    String error;
+    ASTPtr columns_list_raw = tryParseQuery(parser, pos, pos+argument.size(), error, false, "", false, 0, 0);
+    bool parse_structure_with_parser = false;
+    if ((parse_structure_with_parser = columns_list_raw != nullptr))
```
Comment (alexey-milovidov) [inline src/Core/ExternalTable.cpp:79]: Use `if (declaration; condition)` syntax.

File: src/Core/ExternalTable.cpp [hunk 4]
```diff
@@ -55,23 +70,61 @@ void BaseExternalTable::clear()
...
 
 void BaseExternalTable::parseStructureFromTypesField(const std::string & argument)
 {
-    std::vector<std::string> vals;
-    splitInto<' ', ','>(vals, argument, true);
+    /// First try to parse table structure with `ParserTypeList`, this allows user to declare Enum types in the structure
+    ParserTypeList parser;
+    const auto * pos = argument.data();
+    String error;
+    ASTPtr type_list_raw = tryParseQuery(parser, pos, pos+argument.size(), error, false, "", false, 0, 0);
```
Comment (alexey-milovidov) [inline src/Core/ExternalTable.cpp:114]: Missing whitespace.

File: tests/queries/0_stateless/02935_external_table_enum_type.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+http_url="http://${CLICKHOUSE_HOST}:${CLICKHOUSE_PORT_HTTP}/?"
+
+curl -s "${http_url}temp_structure=x+Enum8('foo'%3D1,'bar'%3D2),y+Int" -F "$(printf 'temp='"foo"'\t1');filename=data1" -F "query=SELECT * FROM temp"
+curl -s "${http_url}temp_types=Enum8('foo'%3D1,'bar'%3D2),Int" -F "$(printf 'temp='"bar"'\t2');filename=data1" -F "query=SELECT * FROM temp"
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02935_external_table_enum_type.sh:10]: Also add a test for clickhouse-client with --types and --structure.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/02935_external_table_enum_type.sh:10]: Ok, maybe I can add it?

**Comment by alexey-milovidov** [comment]
Without this PR:

```
$ clickhouse-client --external --file <(echo "Hello, world") --name test --format CSV --structure "x Enum('Hello' = 1, 'world' = 2), y String" --query "SELECT * FROM test"
Code: 36. DB::Exception: Odd number of attributes in section structure: 9. (BAD_ARGUMENTS)
Table №0

$ clickhouse-client --external --file <(echo "Hello, world") --name test --format CSV --structure "x String, y String" --query "SELECT * FROM test"
Hello   world
```

**Comment by alexey-milovidov** [comment]
It works... I will merge, then add a new test.

**Comment by alexey-milovidov** [comment]
Actually, I have to make it in a single PR, because otherwise the build will fail...

**Comment by alexey-milovidov** [comment]
I have added a test and merged it with master - now it is ready.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by canhld94** [comment]
@Alexey thanks for helping me with this PR. Sorry recently I've been too
busy and cannot arrange time to look at upstream PRs.

**Comment by alexey-milovidov** [comment]
Thank you, that's a really good contribution!

Often times, a pull request is ready for 99%, and we have to finish the most cumbersome part, like adding extra tests, or polishing for corner cases.

**Pull request merged**